lib/web/app/transport.go

```diff
@@ -231,41 +225,30 @@ func (t *transport) rewriteRequest(r *http.Request) error {
 
 // DialContext dials and connect to the application service over the reverse
 // tunnel subsystem.
-func (t *transport) DialContext(ctx context.Context, _, _ string) (net.Conn, error) {
-	var err error
-	var conn net.Conn
-
-	t.servers.Range(func(serverID, appServerInterface interface{}) bool {
-		appServer, ok := appServerInterface.(types.AppServer)
-		if !ok {
-			t.c.log.Warnf("Failed to load AppServer, invalid type %T", appServerInterface)
-			return true
-		}
-
-		var dialErr error
-		conn, dialErr = dialAppServer(ctx, t.c.proxyClient, t.c.identity.RouteToApp.ClusterName, appServer)
-		if dialErr != nil {
-			if isReverseTunnelDownError(dialErr) {
-				t.c.log.Warnf("Failed to connect to application server %q: %v.", serverID, dialErr)
-				t.servers.Delete(serverID)
-				// Only goes for the next server if the error returned is a
-				// connection problem. Otherwise, stop iterating over the
-				// servers and return the error.
-				return true
-			}
+func (t *transport) DialContext(ctx context.Context, _, _ string) (conn net.Conn, err error) {
+	servers := t.c.servers[:0]
+	for i, appServer := range t.c.servers {
+		conn, err = dialAppServer(ctx, t.c.proxyClient, t.c.identity.RouteToApp.ClusterName, appServer)
+		if err != nil && isReverseTunnelDownError(err) {
+			t.c.log.Warnf("Failed to connect to application server %q: %v.", appServer, err)
+			// Continue to the next server if there is an issue
+			// establishing a connection because the tunnel is not
+			// healthy. Reset the error to avoid returning it if
+			// this is the last server.
+			err = nil
+			continue
 		}
 
-		// "save" dial error to return as the function error.
-		err = dialErr
-		return false
-	})
+		servers = append(servers, t.c.servers[i:]...)
+		break
+	}
 
-	if err != nil {
-		return nil, trace.Wrap(err)
+	for i := len(servers); i < len(t.c.servers); i++ {
+		t.c.servers[i] = nil
 	}
```

Comment:
Could you add a comment explaining this? I don't quite understand what this is doing.